drivers/ipvlan/ipvlan_endpoint.go

```diff
@@ -0,0 +1,60 @@
+package ipvlan
+
+import (
+	"fmt"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/libnetwork/driverapi"
+	"github.com/docker/libnetwork/netutils"
+	"github.com/vishvananda/netlink"
+)
+
+// CreateEndpoint assigns the mac, ip and endpoint id for the new container
+func (d *driver) CreateEndpoint(nid, eid string, ifInfo driverapi.InterfaceInfo,
+	epOptions map[string]interface{}) error {
+
+	if err := validateID(nid, eid); err != nil {
+		return err
+	}
+	n, err := d.getNetwork(nid)
+	if err != nil {
+		return fmt.Errorf("network id %q not found", nid)
+	}
+	ep := &endpoint{
+		id:   eid,
+		addr: ifInfo.Address(),
+		mac:  ifInfo.MacAddress(),
+	}
+	if ep.addr == nil {
+		return fmt.Errorf("create endpoint was not passed an IP address")
+	}
+	if ep.mac == nil {
+		ep.mac = netutils.GenerateMACFromIP(ep.addr.IP)
+		if err := ifInfo.SetMacAddress(ep.mac); err != nil {
+			return err
+		}
+	}
+	n.addEndpoint(ep)
```

Comment:
Thanks for the thoughts, I totally get your point on isolating them. On the other hand, the ability for disparate ip prefixes to be reachable without an explicit router/gateway is what makes L3 mode interesting by breaking the mold of viewing a network as an ip subnet. 

- Putting up two examples for clarity for folks.

**Example #2:** Creates networks on the same shared parent interface `-o host_iface=eth0.99` where the two hosts can talk to one another. 

- Shared Network Domain - networks sharing the same parent interface `-o host_iface=`

```
### Create one more l3 network using the same parent iface as the previous network with a different subnet
docker network create -d ipvlan --subnet=192.168.91.0/24 -o host_iface=eth0.99 -o ipvlan_mode=l3 backend_net1
docker network create -d ipvlan --subnet=192.168.99.0/24 -o host_iface=eth0.99 -o ipvlan_mode=l3 backend_net2

### Start containers on backend_net1
docker run --net=backend_net1 --name=backend_host --ip=192.168.91.10 -itd alpine /bin/sh

# Do a test ping from the next host to 192.168.99.10 via 192.168.91.10
# *Note* this works across L3 boundaries because you share the 
# same parent interface (eth0.99) even though it is across discontiguous subnets
docker run --net=backend_net2 --ip=192.168.99.10 -it --rm alpine /bin/sh

# Containers on the SAME parent interface (-o host_iface=) CAN ping one another even across L3 boundaries (e.g. subnets)
/ # ping  192.168.91.10
PING 192.168.91.10 (192.168.91.10): 56 data bytes
64 bytes from 192.168.91.10: seq=0 ttl=64 time=0.060 ms
64 bytes from 192.168.91.10: seq=1 ttl=64 time=0.078 ms
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.060/0.069/0.078 ms

```

**Example #2:** creates networks on two different parent interfaces  `-o host_iface=eth0.94` & `-o host_iface=eth0.95` which prevent the slave interfaces in the containers from reaching one another, unless there was an external routing process to hairpin the traffic between containers which is beyond the scope of the driver.


- Isolated Network: networks not sharing the same parent  interface `-o host_iface=`

```
### Same network prefixes but with 
docker network create -d ipvlan --subnet=192.168.94.0/24 -o host_iface=eth0.94 -o ipvlan_mode=l3 web_net
docker network create -d ipvlan --subnet=192.168.95.0/24 -o host_iface=eth0.95 -o ipvlan_mode=l3 db_net

### Start containers on each network
docker run --net=web_net --ip=192.168.94.10 --name=web_test2 -itd alpine /bin/sh


# Containers on DIFFERENT parent interface (-o host_iface=) can NOT ping one another even across L3 boundaries (e.g. subnets)
docker run --net=db_net -it --rm alpine /bin/sh
ping 192.168.95.10
PING 192.168.95.10 (192.168.95.10): 56 data bytes
--- 192.168.95.10 ping statistics ---
2 packets transmitted, 0 packets received, 100% packet loss
```

To add a bit of context for others not familiar with ipvlan l3 mode is assigning each container host routes `/32` for mobility from subnet/l2 locality. L3 needs a route to be distributed throughout a cluster to be functional whether its integrated with the network fabric IGP or in an overlay with a distributed datastore etc. I'm thinking implicitly isolating based on IP subnet may limit some of the intended user/3rd party plugin use cases.